### DISCUSSION TITLE
A few problems regarding using X forwarding to connect to GNOME desktop

### DISCUSSION THREAD

**Discussion by Heng-Zhou** [microsoft/WSL#10898] [Q&A]
Environment:

> WSL version: 2.0.9.0
> Kernel version: 5.15.133.1-1
> WSLg version: 1.0.59
> MSRDC version: 1.2.4677
> Direct3D version: 1.611.1-81528511
> DXCore version: 10.0.25131.1002-220531-1700.rs-onecore-base2-hyp
> Windows version: 10.0.19045.3693 (22H2 Education)
> Ubuntu version: 22.04.3 LTS (GNU/Linux 5.15.133.1-microsoft-standard-WSL2 x86_64)

I installed Ubuntu 22.04 and GNOME desktop in WSL2. Below are the steps:

1) sudo apt update, sudo apt upgrade
2) sudo apt-mark hold acpid acpi-support
3) sudo apt install ubuntu-desktop gnome
4) Add the following 2 lines into ~/.bashrc:

> export DISPLAY=$(cat /etc/resolv.conf | grep nameserver | awk '{print $2}'):0
> export LIBGL_ALWAYS_INDIRECT=1

5) `git clone https://github.com/DamionGans/ubuntu-wsl2-systemd-script.git`
  `cd ubuntu-wsl2-systemd-script/`
  `bash ubuntu-wsl2-systemd-script.sh`
6) exit Ubuntu and in Windows command line run `wsl --shutdown`
7) Start VcXsrv and in the launcher, uncheck "Native opengl" and check "Disable access control"
8) Enter wsl

The problems are:

1) If I run `gnome-session`, the GNOME desktop appears but crashes after about one minute. 
![image](https://github.com/microsoft/WSL/assets/111557423/6e6d7a16-2b81-4eae-ac24-235e365befdb)
If I click "Log Out" button and run `gnome-session` again in WSL, I receive the following error:

> ** (process:970): WARNING **: 07:43:49.894: Could not make bus activated clients aware of XDG_CURRENT_DESKTOP=GNOME environment variable: Could not connect: No such file or directory
> Terminated

If I run `export XDG_CURRENT_DESKTOP=ubuntu:GNOME` and run `gnome-session` again, I receive just "Terminated".

2) If I run `dbus-launch --exit-with-session gnome-session`, the GNOME desktop does not crash, but I get a lot of other problems. 

2.1) dialog freezes easily on desktop, and I can find no way to remove them.
![image](https://github.com/microsoft/WSL/assets/111557423/65589a4c-2ded-40ad-82f4-c0f26eb3487f)

2.2) Because of 2.1), the frozen dialog covers the desktop, so I can't log out or power down the GNOME desktop using its GUI. If I press Ctrl-C in command line and try to run `dbus-launch --exit-with-session gnome-session` again, I receive an error saying 
> dbus[1560]: Unable to set up transient service directory: XDG_RUNTIME_DIR "/run/user/1000" not available: No such file or directory

2.3) If I run `dbus-launch --exit-with-session gnome-session` with sudo, I can go back into the GNOME desktop, but the desktop is still covered by frozen dialogs, and I cannot type in the search bar of Activities.

2.4) If I click "Power Off" in the GNOME desktop, the desktop does not power off, but stops response unless I press Ctrl-C in WSL command line. In that case, the desktop shows this:
![image](https://github.com/microsoft/WSL/assets/111557423/302e9f97-54b0-4cfc-b021-20867d01def5)
and the erroneous desktop quits if I click "Log Out" button. If I quit WSL, run `wsl --shutdown` in Windows command line and run wsl again, `dbus-launch --exit-with-session gnome-session` can start the GNOME desktop but got the same problem of 2.3), and so on and on ...

Do you know how these problems are caused according to my steps, or how to fix these issues, or how to reset the GNOME desktop to remove the frozen dialogs (if I don't have to redo the whole thing due to brick)? Thanks a lot.

**Comment by Yeee3-pei** [comment]
![螢幕擷取畫面 2023-12-31 223432](https://github.com/microsoft/WSL/assets/114487183/aec61c3a-a01d-47b5-979e-38272605ad43)
Is there any solve of this problem?

**Reply by Heng-Zhou** [reply]
Please refer to my answer: https://gist.github.com/Ta180m/e1471413f62e3ed94e72001d42e77e22?permalink_comment_id=4793852#gistcomment-4793852

Happy new year.

**Reply by Yeee3-pei** [reply]
got it! thx

**Reply by Nitjsefnie** [reply]
Hello, I just came across exactly the same issue, however the link seems to no longer work. What was the solution?

**Reply by LukasSkywalker** [reply]
This was luckily archived by the Wayback Machine. Here's the content from May 19, 2024:

> [!NOTE]  
> From https://web.archive.org/web/20240519071650/https://gist.github.com/Ta180m/e1471413f62e3ed94e72001d42e77e22

# WSL 2 GNOME Desktop

**NOTE:** If you want the ultimate Linux desktop experience, I highly recommend installing Linux as your main OS. I no longer use Windows (except in a VM) so I will not be maintaining this guide anymore.

Think Xfce looks dated? Want a conventional Ubuntu experience? This tutorial will guide you through installing Ubuntu's default desktop environment, GNOME.

GNOME is one of the more complex — and that means more difficult to run — desktop environments, so [for](https://www.reddit.com/r/Windows10/comments/8dnyig/using_a_gui_with_wsl_on_windows_10_ubuntu_wsl/) [years](https://www.reddit.com/r/bashonubuntuonwindows/comments/9amqjr/gnome_desktop_using_vcxsrv_cant_make_it_work/) [people](https://askubuntu.com/questions/1169695/gnome-under-ubuntu-18-04-for-windows-wont-run) [couldn't](https://superuser.com/questions/1559755/gnome-desktop-environment-on-ubuntu-wsl-starting-error) [figure](https://www.reddit.com/r/bashonubuntuonwindows/comments/haof3n/cant_run_gnome_in_ubuntu_2004_lts_wsl/) [out](https://www.reddit.com/r/bashonubuntuonwindows/comments/eqwoxy/gnome_and_kde_in_wsl/) [how](https://www.reddit.com/r/bashonubuntuonwindows/comments/8dp2yu/gnome_or_kde_on_wsl/) [to](https://www.reddit.com/r/bashonubuntuonwindows/comments/hh0q9s/gnomesession_wsl2/) [run](https://www.reddit.com/r/bashonubuntuonwindows/comments/b222hp/running_everything_on_gnome/) [it](https://www.reddit.com/r/bashonubuntuonwindows/comments/8yfw8z/gnome_on_windows/) on WSL 2. On [WSL 1](https://github.com/Microsoft/WSL/issues/637) it could only run using very complicated methods that didn't transfer to well WSL 2. Any forlorn attempts to run it on WSL 2 only resulted in a smoldering heap of error messages.

__But now you can!__


## Requirements:
 - WSL 2
 - Ubuntu 20.04 (other distros not tested)
 - An X server for Windows, such as [VcXsrv](https://sourceforge.net/projects/vcxsrv/)
 - Basic knowledege on how to run GUI apps with WSL 2 (not required but highly recommended)


## Getting ready

You've been regularly updating your distro, haven't you?

```sh
sudo apt update
sudo apt upgrade
```

Install GNOME: (maybe go eat a snack while it's installing?)

```sh
sudo apt install ubuntu-desktop gnome
```

Open up your `~/.bashrc`:

```sh
nano ~/.bashrc
```

And paste this in at the end and save:

```sh
export DISPLAY=$(cat /etc/resolv.conf | grep nameserver | awk '{print $2}'):0
export LIBGL_ALWAYS_INDIRECT=1
```

If you try to start GNOME now, you'll get a lot of errors. Something along the lines of this, but a ton more errors:

```sh
Unable to init server: Could not connect: Connection refused

(gnome-session-check-accelerated:6054): Gtk-WARNING **: 11:04:51.973: cannot open display: :0
Unable to init server: Could not connect: Connection refused

(gnome-session-check-accelerated:6055): Gtk-WARNING **: 11:04:52.234: cannot open display: :0
gnome-session-binary[6044]: WARNING: software acceleration check failed: Child process exited with code 1
gnome-session-binary[6044]: CRITICAL: We failed, but the fail whale is dead. Sorry....
```

The trick is to enable `systemd`: (note that this does break a lot of stuff such as Visual Studio Code Remote)

```sh
git clone https://github.com/DamionGans/ubuntu-wsl2-systemd-script.git
cd ubuntu-wsl2-systemd-script/
bash ubuntu-wsl2-systemd-script.sh
```

Now shut down WSL 2: (run this in Windows)

```sh
wsl --shutdown
```


## Starting GNOME

First, fire up your X server on Windows. Make sure you let it through your firewall and disable access control.

Now, start up Ubuntu again and start GNOME:

```sh
gnome-session
```

If you don't get any error messages, you should be good. Wait a few seconds for GNOME to start up.

<img width="1007" alt="desktop" src="https://user-images.githubusercontent.com/50341827/88995077-68591600-d2af-11ea-91e6-55d528a57dd2.png">

Now you have a great GUI desktop and you won't need any intensive virtual machines anymore!

Profit?


## Notes

 - You can disable the screensaver with `gsettings set org.gnome.desktop.session idle-delay 0`.
 - You can also try KDE Plamsa using a similar method! Just `sudo apt install kde-plasma-desktop` instead and start it with `startplasma-x11`.


## Troubleshooting

If you can't get this to work, try [Xfce](https://github.com/QMonkey/wsl-tutorial/blob/master/README.wsl2.md).

If you still can't get it to work, you can ask for help on an online forum such as [r/bashonubuntuonwindows](https://www.reddit.com/r/bashonubuntuonwindows/).
